### DISCUSSION TITLE
Enable more optimizations for the compiler like LTO, PGO, PLO

### DISCUSSION THREAD

**Discussion by zamazan4ik** [shader-slang/slang#5632] [Ideas]
Hi!

Here I want to propose additional optimizations that can be enabled for the Slang compiler (and maybe other related Slang tools too).

### Link-Time Optimization (LTO)

Link-Time Optimization (LTO) is an additional compiler optimization that gives more context to the compiler during the optimization process which leads to better optimization results. It's supported by all major C++ compilers like GCC, Clang, MSVC (`-flto` for GCC e.g.). It also could be enabled on the CMake level via `set(CMAKE_INTERPROCEDURAL_OPTIMIZATION TRUE)`.

I tried to search over the build scripts but didn't find a place where LTO is enabled so I suppose it's disabled for now.

### Profile-Guided Optimization (PGO) and Post-Link Optimization (PLO)

According to my [benchmarks](https://github.com/zamazan4ik/awesome-pgo), PGO measurably helps to optimize various applications (its CPU efficiency more precisely) in several domains, especially compilers, static analysis, linters, etc - see Clang, Rustc, clang-tidy, Rust Analyzer and many other examples. That's why I think applying PGO (and a similar PLO technique via [LLVM BOLT](https://github.com/llvm/llvm-project/blob/main/bolt/README.md)) could help to optimize Slang further even after LTO.

Thank you.

**Comment by jkwak-work** [comment]
I don't think the techniques mentioned here make sense for Slang.

For shaders, there is no known way to collect the "profile" data where there can be thousands of pixels running at the same time every seconds, if not milliseconds.

The linking is also not very relevant for shader, because each compilation units for shaders are not really linked in a way that cpp object files are linked.

**Reply by zamazan4ik** [reply]
> For shaders, there is no known way to collect the "profile" data where there can be thousands of pixels running at the same time every seconds, if not milliseconds.

The PGO profile is collected from shaders itself but from different parts of Slang, that are executed during the shader compilation. My proposal is about optimizing with PGO Slang itself, not the resulting shader programs.

> The linking is also not very relevant for shader, because each compilation units for shaders are not really linked in a way that cpp object files are linked.

I am talking about optimizing the Slang compiler itself with LLVM BOLT, not shader-based programs.

**Reply by jkwak-work** [reply]
I apologize for my misunderstanding.
It makes sense if we are talking about the performance of Slang itself.

**Reply by csyonghe** [reply]
I thought LTO is enabled by default on high optimization levels, but if it is not it is definitely worth enabling.

**Reply by jkwak-work** [reply]
I have a PR to enable LTCG for release builds.
https://github.com/shader-slang/slang/pull/5678
